### PR TITLE
[feature] Low latency mode

### PR DESCRIPTION
**Pull request by tylerjw**
### Description

This change allows the user to tie the output of servo to the input.  By using a condition variable in this new `low_latency_mode` servo will run each time it gets input.

Includes a test that validates this works as expected.

### Checklist
- [x] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [x] Create tests, which fail without this PR [reference](https://ros-planning.github.io/moveit_tutorials/doc/tests/tests_tutorial.html)
- [x] While waiting for someone to review your request, please help review [another open pull request](https://github.com/ros-planning/moveit/pulls) to support the maintainers

### CODE REVIEW

File: moveit_ros/moveit_servo/src/servo_calcs.cpp
Comment (AndyZe) [reply moveit_ros/moveit_servo/src/servo_calcs.cpp:209]: We should be able to resume the main ServoCalcs() loop easily

Comment (AndyZe) [reply moveit_ros/moveit_servo/src/servo_calcs.cpp:209]: Also, doesn't `main()` have kind of a special meaning in C++? Not sure this function should be named so

**Review by AndyZe — COMMENTED** [review]
Overall, I'm not sure I like this. Say the incoming command rate is 250 Hz (4 ms). If I understand, this method might save 3 ms of latency on the first command. But the long-term average latency is still going to be 4 ms. So there's very little gain but more complexity and mutexes involved.

**Comment by AndyZe** [comment]
I'm wondering if this could be simpler:

`std::atomic_bool_ got_first_command_ = false;`

```
void ServoCalcs::twistStampedCB(const geometry_msgs::TwistStampedConstPtr& msg)
{
  ...

  // notify that we have a new input
 got_first_command_ = true;
}
```

Reset got_first_command_ to false every time a stop is requested

File: moveit_ros/moveit_servo/src/servo_calcs.cpp
Comment (tylerjw) [reply moveit_ros/moveit_servo/src/servo_calcs.cpp:209]: main doesn't have a special meaning when it is part of a class... it is a pattern i've seen to use main as the function you start a thread with.

If this function returns it can be joined (the thread is finished and could be joined) which will happen if someone calls start again.

**Comment by tylerjw** [comment]
`condition_variable` is a stupid name for what it does.  It makes it easier for me to think of it as a mutex protected signal you can send from one thread to unblock another.  What this does is in `low_latency_mode` the main loop blocks and resumes imidiatly when it receives a new command.  If servo was previously running at 50Hz with another loop commanding it (like the pose tracker) at 50Hz turning this mode on would eliminate 20ms of input delay.  This mode is useful if it is being sent commands by another high rate loop as it just ties the rate of servo to that loop without any of the input latency.

> I'm wondering if this could be simpler:
> 
> `std::atomic_bool_ got_first_command_ = false;`
> 
> ```
> void ServoCalcs::twistStampedCB(const geometry_msgs::TwistStampedConstPtr& msg)
> {
>   ...
> 
>   // notify that we have a new input
>  got_first_command_ = true;
> }
> ```
> 
> Reset got_first_command_ to false every time a stop is requested

I don't understanding what you are suggesting with a `got_first_command_` bool.  What I'm trying to achieve here is to have servo unblock and calculate an output imidiatly after receiving a new input while still preserving the thread separation and nice ros interface it already has.

> Say the incoming command rate is 250 Hz (4 ms) and the controller accepts commands at 100 Hz (10 ms). If I understand, this method might save 3 ms of latency on the first command. But the long-term average command rate to the controller is still going to be 10 ms. So there's very little gain but more complexity.

I am probably misunderstanding something but I don't think it would be useful to run servo at a higher rate than the ros controller accepts commands as you would be calculating commands that would just be dropped at best.  Generally wouldn't you want your robot controller to be running at a higher (or the same) rate than servo or am I misunderstanding something?  Similarly if you are feeding servo commands at a higher rate than it is running roughly half of your commands will be dropped.  That also seems like it could have non-linear affects that would be generally bad.

File: moveit_ros/moveit_servo/src/servo_calcs.cpp
Comment (AndyZe) [inline moveit_ros/moveit_servo/src/servo_calcs.cpp:203]: This overlaps with the existing functionality, `setPaused()`

**Comment by AndyZe** [comment]
It looks like it takes about 50 to a few hundred milliseconds to start a thread. Sorry Tyler, I'm not digging it

https://stackoverflow.com/questions/18274217/how-long-does-thread-creation-and-termination-take-under-windows

**Comment by AndyZe** [comment]
> I don't understand what you are suggesting with a `got_first_command_` bool.

I'll expand my comment

File: moveit_ros/moveit_servo/src/servo_calcs.cpp
Comment (tylerjw) [reply moveit_ros/moveit_servo/src/servo_calcs.cpp:203]: It is a private function and not part of the exposed interface.  I created this function because I needed the block of code in the destructor and in start().

**Comment by AndyZe** [comment]
On the bright side, one issue this PR brought to mind is that we should decrease this period just a bit in ServoCalcs so there's always a fresh command waiting to be published:

`  timer_ = nh_.createTimer(period_, &ServoCalcs::run, this);`

Should be something like:

`  timer_ = nh_.createTimer(**0.9 *** period_, &ServoCalcs::run, this);`

**Comment by tylerjw** [comment]
> It looks like it takes about 5 to a few hundred milliseconds to start a thread. Sorry Tyler, I'm not digging it
> 
> https://stackoverflow.com/questions/18274217/how-long-does-thread-creation-and-termination-take-under-windows

Yes, we are already eating that cost with the AsyncSpinner.  I don't see how that matters though.  That time is spent in the start() method not in the loop.

**Comment by tylerjw** [comment]
> On the bright side, one issue this PR brought to mind is that we should decrease this period just a bit in ServoCalcs so there's always a fresh command waiting to be published:
> 
> ` timer_ = nh_.createTimer(period_, &ServoCalcs::run, this);`
> 
> Should be something like:
> 
> ` timer_ = nh_.createTimer(0.9 * period_, &ServoCalcs::run, this);`

The user could just decrease the period in the config by 10% if that is what they wanted.  That will reduce the latency by 10% as that would just increase the frequency of the loop by 10%.  That is completely different from using a condition variable to tie running servo calcs to the input.

**Comment by AndyZe** [comment]
> I don't understand what you are suggesting with a `got_first_command_` bool

You're right, this is hard to do now that the work is started by a timer:   `timer_ = nh_.createTimer(period_, &ServoCalcs::run, this);`

But it used to be done in a while loop, and then it would have been a lot easier:

```
while (// jog loop conditions)
{
  // Wait for first commands
  while(ros::ok() && !got_first_command_)
  {
    ros::Duration(0.001).sleep():
  }

  // Do jog calculations here

  ros::Duration(0.001).sleep():
}
```

**Comment by AndyZe** [comment]
> The user could just decrease the period in the config by 10% if that is what they wanted. That will reduce the latency by 10% as that would just increase the frequency of the loop by 10%. That is completely different from using a condition variable to tie running servo calcs to the input.

Well, that's not what I'm talking about. We'll have to sync offline, I guess

**Comment by tylerjw** [comment]
> But it used to be done in a while loop, and then it would have been a lot easier:
> 
> ```
> while (// jog loop conditions)
> {
>   // Wait for first command
>   while(ros::ok() && !got_first_command_)
>   {
>     ros::Duration(0.001).sleep():
>   }
> 
>   // Do jog calculations here
> 
>   ros::Duration(0.001).sleep():
> }
> ```

The condition variable does this without having to poll a variable.  The condition variable triggers the continuation of the calc thread as soon as the input callback returns (without tying up a thread just polling on a bool like this would do).

The advantage of using a timer is it doesn't create "yet another thread" and allows you to more optimally use your cpu in a external to servo way.  There is no way to trigger the running of a timer in this way.  One could put the calcuations in the callback but that will have a negative affect on the system of making everything in the process less responsive to new ros messages received over the network because one of the closures for servicing the callbacks will now take a relatively long time to return.  To me using a thread and condition variable seems like a nice compromise.

One thing to note is that this PR does not externally change servo for existing use cases where someone doesn't drive it with another high rate (or similar rate) loop.

**Comment by AndyZe** [comment]
Can we take advantage of `timer_.start()` and `timer_.stop()` to do this in a less disruptive way?

**Comment by tylerjw** [comment]
> Can we take advantage of `timer_.start()` and `timer_.stop()` to do this in a less disruptive way?

I don't think there is any guarentee that when you call start() it imidiatly runs.  It seems to me that using a timer for a single run on a separate thread is counter to what the timer is for.  The way I'm achieving that is to block the calc thread and unblock it in the callback.

This is not disrutive to the external user of servo and I think a condition variable is a good solution.  The largest downside to condition variables is they are not intuitive to use (I have to read the documentation and write tests to validate my understanding each time I use them).

File: moveit_ros/moveit_servo/src/servo_calcs.cpp
Comment (jliukkonen) [inline moveit_ros/moveit_servo/src/servo_calcs.cpp:245]: Nit: set both the start time and run duration as const.

File: moveit_ros/moveit_servo/src/servo_calcs.cpp
Comment (jliukkonen) [inline moveit_ros/moveit_servo/src/servo_calcs.cpp:235]: This is something I don't really know an answer to, but I assume the author does and curious to hear the answer. So, wouldn't this be affected by spurious wakeups? And would it be better to add a check for, e.g., a flag that would be set by either the `twistStampedCB` and `jointCmdCB` callback functions? That is, add a predicate to the `.wait()` method?

**Review by jliukkonen — COMMENTED** [review]
Implementation and the idea both look very good to me. Left couple comments and after clearing those, I'm okay with merging this.

File: moveit_ros/moveit_servo/src/servo_calcs.cpp
Comment (tylerjw) [reply moveit_ros/moveit_servo/src/servo_calcs.cpp:235]: Yes, this could be triggered by a spurious wakeup.  I can add a predicate to this.

**Review by jliukkonen — APPROVED** [review]

File: moveit_ros/moveit_servo/src/servo_calcs.cpp
Comment (tylerjw) [reply moveit_ros/moveit_servo/src/servo_calcs.cpp]: I added the predicate.  It now is not susceptible to the spurious wakeups a condition variable can be.

File: moveit_ros/moveit_servo/include/moveit_servo/servo_calcs.h
Comment (AndyZe) [inline moveit_ros/moveit_servo/include/moveit_servo/servo_calcs.h:108]: Let's improve these comments and rename the functions:

  /** \brief Run the main calculation loop */
  void mainCalcLoop();

/** \brief Do calculations for a single iteration. Publishes one outgoing command */
void calculateSingleIteration();

**Review by AndyZe — COMMENTED** [review]
Here's some fresh feedback after letting this simmer awhile. I still think this functionality is borderline not worthwhile (we're talking about time savings of maybe 2 ms for a typical scenario).

Let's get rid of the yaml parameter and always use "low latency mode" by default, since I can't think of a case where we wouldn't want to use it.

Also, delete the new test since it's not a new mode any more -- it will just be the default mode.

I still need to test it, too

**Comment by tylerjw** [comment]
I'm sorry I didn't include this with the PR but I wrote some profiling code and here is the result from a modified version of the cpp_interface_test to publish the time between when servo receives a command (the callbacks) to when it publishes based on that input:

low_latency_v_normal

If you want the profiliing code message me, it's an annoyingly long diff.

**Review by AndyZe — APPROVED** [review]
Made a cleanup commit. Looks good to me now.

**Pull request merged**
